File: frontend/src/api/coding.ts
import apiClient from "./client";

export interface CodingProblem {
  id: number;
  title: string;
  slug: string;
  difficulty: "Easy" | "Medium" | "Hard";
  category: string;
  description: string;
  constraints?: string;
  examples: Array<{ input: string; output: string; explanation?: string }>;
  starter_code: Record<string, string>;
  points: number;
  is_solved?: boolean;
}

export interface TestCaseResult {
  test_case: number;
  input: string;
  expected_output: string;
  actual_output: string;
  passed: boolean;
  execution_time_ms: number;
  error?: string;
}

export interface CodingRunResponse {
  passed: boolean;
  passed_count: number;
  total_count: number;
  test_results: TestCaseResult[];
  runtime_ms: number;
  memory_mb: number;
  error?: string;
}

export interface CodingSubmitResponse {
  submission_id: number;
  status: string;
  passed_count: number;
  total_count: number;
  runtime_ms: number;
  memory_mb: number;
  xp_awarded: number;
  error_message?: string;
}

export interface UserCodingStats {
  solved_count: number;
  total_count: number;
  easy_solved: number;
  medium_solved: number;
  hard_solved: number;
  streak: number;
  total_xp: number;
  completion_percentage: number;
}

export const codingApi = {
  getProblems: (params?: { category?: string; difficulty?: string; search?: string; status_filter?: string }) =>
    apiClient.get<CodingProblem[]>("/coding/problems", { params }),

  getProblem: (id: number) =>
    apiClient.get<CodingProblem>(`/coding/problems/${id}`),

  runCode: (id: number, data: { language: string; code: string }) =>
    apiClient.post<CodingRunResponse>(`/coding/problems/${id}/run`, data),

  submitCode: (id: number, data: { language: string; code: string }) =>
    apiClient.post<CodingSubmitResponse>(`/coding/problems/${id}/submit`, data),

  explainWithAI: (id: number, data: { language?: string; code?: string; mode?: string }) =>
    apiClient.post<{ explanation: string }>(`/coding/problems/${id}/explain-ai`, data),

  getUserStats: () =>
    apiClient.get<UserCodingStats>("/coding/user-stats"),
};

File: backend/app/schemas/coding.py
from typing import Any, Dict, List, Optional
from pydantic import BaseModel


class ProblemExample(BaseModel):
    input: str
    output: str
    explanation: Optional[str] = None


class TestCaseItem(BaseModel):
    input: str
    expected_output: str
    is_hidden: Optional[bool] = False


class StarterCode(BaseModel):
    python: str
    cpp: str
    java: str
    javascript: str


class CodingProblemOut(BaseModel):
    id: int
    title: str
    slug: str
    difficulty: str  # Easy, Medium, Hard
    category: str
    description: str
    constraints: Optional[str] = None
    examples: List[Dict[str, Any]]
    starter_code: Dict[str, str]
    points: int
    is_solved: Optional[bool] = False

    class Config:
        from_attributes = True


class TestCaseResult(BaseModel):
    test_case: int
    input: str
    expected_output: str
    actual_output: str
    passed: bool
    execution_time_ms: float
    error: Optional[str] = None


class CodingRunRequest(BaseModel):
    language: str  # python, cpp, java, javascript
    code: str


class CodingRunResponse(BaseModel):
    passed: bool
    passed_count: int
    total_count: int
    test_results: List[TestCaseResult]
    runtime_ms: float
    memory_mb: float
    error: Optional[str] = None


class CodingSubmitRequest(BaseModel):
    language: str
    code: str


class CodingSubmitResponse(BaseModel):
    submission_id: int
    status: str  # Accepted, Wrong Answer, etc.
    passed_count: int
    total_count: int
    runtime_ms: float
    memory_mb: float
    xp_awarded: int
    error_message: Optional[str] = None


class AIExplainRequest(BaseModel):
    language: Optional[str] = "python"
    code: Optional[str] = None


class AIExplainResponse(BaseModel):
    explanation: str


class UserCodingStats(BaseModel):
    solved_count: int
    total_count: int
    easy_solved: int
    medium_solved: int
    hard_solved: int
    streak: int
    total_xp: int
    completion_percentage: float
